R/tags.R

```diff
@@ -309,19 +311,10 @@ tagHasAttribute <- function(tag, attr) {
 #' @export
 tagGetAttribute <- function(tag, attr) {
   throw_if_tag_function(tag)
-  # Find out which positions in the attributes list correspond to the given attr
-  attribs <- tag$attribs
-  attrIdx <- which(attr == names(attribs))
-
-  if (length(attrIdx) == 0) {
-    return (NULL)
-  }
-
-  # Convert all attribs to chars explicitly; prevents us from messing up factors
-  result <- lapply(attribs[attrIdx], as.character)
-  # Separate multiple attributes with the same name
-  result <- paste(result, collapse  = " ")
-  result
+  # Must flatten all tag attribs to handle legacy situations
+  #   where there are two entries for the same name
+  attribs <- flattenTagAttribs(tag$attribs)
+  attribs[[attr]]
```

Comment:
This'll end up calling `as.character()` on _all_ the attributes now, right? Seems like we should avoid doing that (I think using `as.character()` here  was a mistake in the 1st place)

Reply:
Should we take the stance that all values should be kept in an array?

So if it is a character and if we detect a space, then split the value into parts.  Otherwise, only collect the values.

Then at print time, we can convert to character and paste together.

Reply:
Yeah, I think it does make sense to always store `"a b"` as `c("a", "b")`. However, we have to be careful about handling other types -- for example, lists, factors, and even tag objects.

See https://github.com/rstudio/shiny/pull/3353 for an example of where this was an issue.

Maybe we need one code path for handling the common case, where all instances of a particular attribute are character vectors (as in `div(a = "x", a = "y")`).

But if any instances of an attribute are a different type, just leave them untouched. It might make sense to also warn if there are multiple instances of an attribute where at least one is not a char vector, as in `div(a = "x", a = span("hello"))`.

Reply:
See https://github.com/rstudio/htmltools/pull/212#issuecomment-823559692